File: rxpermission/src/main/java/com/vanniktech/rxpermission/RxPermission.java
package com.vanniktech.rxpermission;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.annotations.CheckReturnValue;
import io.reactivex.rxjava3.annotations.NonNull;

public interface RxPermission {
  /** Requests a single permission. */
  @NonNull @CheckReturnValue Single<Permission> request(@NonNull String permission);

  /** Requests multiple permissions. */
  @NonNull @CheckReturnValue Observable<Permission> requestEach(@NonNull String... permissions);

  /** Returns true when the given permission is granted. */
  @CheckReturnValue boolean isGranted(@NonNull String permission);

  /** Returns true when the given permission is revoked by a policy. */
  @CheckReturnValue boolean isRevokedByPolicy(@NonNull String permission);
}

File: rxpermission/src/main/java/com/vanniktech/rxpermission/RxPermission.kt
package com.vanniktech.rxpermission

import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.annotations.CheckReturnValue

interface RxPermission {
  /** Requests a single permission.  */
  @CheckReturnValue fun request(permission: String): Single<Permission>

  /** Requests multiple permissions.  */
  @CheckReturnValue fun requestEach(vararg permissions: String): Observable<Permission>

  /** Returns true when the given permission is granted.  */
  @CheckReturnValue fun isGranted(permission: String): Boolean

  /** Returns true when the given permission is revoked by a policy.  */
  @CheckReturnValue fun isRevokedByPolicy(permission: String): Boolean

  /** Returns tue when the given permission has been request at least once using either [request] or [requestEach].  */
  @CheckReturnValue fun hasRequested(permission: String): Boolean
}
